TF OD API SSD and FRCNN support

ArutyunovG opened the pull request.

ArutyunovG commented:
Reference #13896 #14697

ArutyunovG commented:
@pytorchbot retest this please

ArutyunovG commented:
@pytorchbot retest this please

ArutyunovG commented:
@pytorchbot retest this please

ArutyunovG commented:
@BIT-silence @houseroad @newstzpz @orionr @ezyang 
Also cc @ir413 since this can be interesting for Detectron development.

Hi,
This pull request adds support of

1. SSD models trained in TF OD API or Caffe;
2. Faster R-CNN models trained in TF OD API.

Please, consider reviewing it.

Since changes are quite big, two converted models are added to simplify giving this changes a try and seeing how they work.
This is the [link to download them](https://www.dropbox.com/s/l9cudg1n54sswfo/ConvertedModels.tar.gz?dl=0).

FRCNN is faster_rcnn_resnet50_coco_2018_01_28 trained in COCO.
SSD is VGG_VOC0712_SSD_300x300 trained in Caffe, i.e. from original SSD paper.

Both models were tested for exact identity of inference with the results, obtained in their original frameworks.

Achieving this identity required changing of operators: 

BBoxTransform
BoxWithNMSLimit
GenerateProposals

as well as adding two new ops:

PriorBox (for SSDs)
CropAndResize (for Faster R-CNN)

All changes are made in a manner to support backward compatibility of existing models/code. Since tests here pass, it seems this goal is achieved, but please let me know, if I missed some details that are not covered by CI.

Much appreciated.

ezyang commented:
cc @fmassa

ArutyunovG commented:
Guys, I added a repo https://github.com/ArutyunovG/pytorch_ssd_pr_helper which can simplify giving these changes a try. 
Assuming, you have built this PR's caffe2 branch and have opencv-python installed.

Commands to give these changes a try are the following

```
git clone https://github.com/ArutyunovG/pytorch_ssd_pr_helper.git
cd pytorch_ssd_pr_helper
./demo.sh
```
Basically, demo script downloads the models, presented above, and runs their inference. 
Results are saved as two png images. One for each model.
So you can see, that models actually make sense.

ArutyunovG commented:
@ezyang @fmassa @BIT-silence @orionr @houseroad 
Ok, with this new repo I think it is very easy for the reviewer to at least try the changes out.
If there is anything else what I can do to simplify review, please let me know.
Meanwhile I'm moving on and will fork from this thread and add more features there separately. At the moment I'm particularly interested in making production protobuf-based models of keypoint Mask R-CNN detectors in Caffe2 from Detectron's python code. Afterwards, either YOLOs (v2 and v3) or DepthwiseConv2DNative from TensorFlow, not sure at the moment.
Best :)

soumith commented:
hey @ArutyunovG 

I spoke to @dzhulgakov last night to figure out what to do with PRs to the caffe2 frontend / backend -- his team has worked on the caffe2 bits in the past.
Currently, as it stands, no one is committing to maintaining that part of the code, it's on life support.
So, getting someone to review this and maintaining it forward will be a struggle.
I'll let @dzhulgakov reply if I missed any information.

ArutyunovG commented:
Hi @soumith 

Thanks for your reply. It's a pity to hear that caffe2 will no longer be developed, since it is a nice framework for production deployment. 

Then I'll continue the development separately from the master branch. In case someone becomes interested, please let me know.

Best wishes and have a nice day!

BIT-silence commented on caffe2/operators/crop_and_resize_op.cc:
```diff
@@ -0,0 +1,366 @@
+                    continue;
+                }
+
+                if (method_ == std::string("bilinear")) {
```
It's fine here, but I'd suggest to make method_ as an instance of a enum class.

ArutyunovG replied:
Switched it to regular `enum` if it's ok.
The reason why not exactly `enum class` is because there seems to be a need to overload output operators for `enum class` for log checks.

BIT-silence commented on caffe2/operators/crop_and_resize_op.h:
```diff
@@ -0,0 +1,94 @@
+#ifndef CROP_AND_RESIZE_OP_H_
```
```suggestion
#ifndef CAFFE2_OPERATORS_CROP_AND_RESIZE_OP_H_
```

ArutyunovG replied:
Fixed.

BIT-silence commented on caffe2/operators/crop_and_resize_op.h:
```diff
@@ -0,0 +1,94 @@
+#ifndef CROP_AND_RESIZE_OP_H_
+#define CROP_AND_RESIZE_OP_H_
```
ditto

ArutyunovG replied:
Fixed.

BIT-silence commented on caffe2/operators/prior_box_op.h:
```diff
@@ -0,0 +1,79 @@
+#ifndef PRIOR_BOX_OP_H_
```
```suggestion
#ifndef CAFFE2_OPERATORS_PRIOR_BOX_OP_H_
```

ArutyunovG replied:
Fixed.

BIT-silence commented on caffe2/operators/bbox_transform_op.cc:
```diff
@@ -110,8 +118,10 @@ bool BBoxTransformOp<float, CPUContext>::RunOnDevice() {
-    CAFFE_ENFORCE_EQ(batch_size, 1);
-    num_rois_per_batch[0] = N;
+    CAFFE_ENFORCE(batch_size == 1 || share_location_);
+    for (int i = 0; i < batch_size; ++i) {
```
```suggestion
    std::fill(num_rois_per_batch, num_rois_per_batch + batch_size, N);
```

ArutyunovG replied:
Fixed.

BIT-silence commented on caffe2/operators/generate_proposals_op.cu:
```diff
@@ -80,37 +95,38 @@ __global__ void GeneratePreNMSUprightBoxesKernel(
-    y1 = fmax(fmin(y1, img_height - 1.0f), 0.0f);
-    x2 = fmax(fmin(x2, img_width - 1.0f), 0.0f);
-    y2 = fmax(fmin(y2, img_height - 1.0f), 0.0f);
+    x1 = fmax(fmin(x1, img_width - im_info_offset), 0.0f);
```
for float32 values, use fminf and fmaxf

ArutyunovG replied:
Fixed here and otherwhere in the file generate_proposals_op.cu.

dzhulgakov commented:
@ArutyunovG - just to clarify, in general Caffe2 backend IS being maintained (and many implementations are being fused with PyTorch frontend APIs).

The comment about "that part of the code, it's on life support." was referring to these particular operators which we were looking for the right reviewer for. @BIT-silence kindly agreed to do review for this PR.

ArutyunovG commented:
@dzhulgakov 

Thank you for the clarification. Nice to know, Caffe2 is not going to be abandoned in general.
Started an issue #19726 to discuss better mechanisms for the future cooperation with the team.

ArutyunovG commented on caffe2/operators/crop_and_resize_op.cc:
```diff
@@ -0,0 +1,366 @@
+
+namespace {
+
+float _roundf(float x)
```
It was causing build error in some of the platforms. Android was there, if I remember correctly. 
Switched it to static_cast<int>(coord_value_float + 0.5f)  as in .cu implementation, if it's ok.

ArutyunovG commented on caffe2/operators/crop_and_resize_op.cu:
```diff
@@ -0,0 +1,393 @@
+
+            const float dtop = (1 - y_lerp) * dY_data[out_idx];
+
+            gpu_atomic_add(static_cast<T>((1 - x_lerp) * dtop),
```
It seems we have to change thread parallelization mechanics to avoid atomic operations when accumulating gradients, as `CUDA_1D_KERNEL_LOOP` asks us to be accurate when adding in the same place from parallel threads.
This implementation is typical for backward pass of operators with semantics "take feature maps and boxes and crop the regions of interests". For example, this is gpu_atomic_add in [roi_align](https://github.com/pytorch/pytorch/blob/6c7135decba2e9090fd663f6d520311a10c84ed8/caffe2/operators/roi_align_gradient_op.cu#L174) (roi align is perhaps semantically closest to crop_and_resize, so I've chosen it).

ArutyunovG commented:
@BIT-silence 

To merge with master and fix demo repo added some updates.

1. For `box_with_nms_limit_op` a285cbcccf4ef7baa297579f5af8e9e2fc5511ae adds new parameter `cls_agnostic_bbox_reg`. 
It is doing the same, what I did with `share_location` for that op! :) So I removed share location and merged the changes in this op and tests.
2. These updates and changes, done for crop_and_resize review fixes, have broken the demo repo. So I updated it and now it works again. 
The commands to run the models remain the same for the user:
```
git clone https://github.com/ArutyunovG/pytorch_ssd_pr_helper.git
cd pytorch_ssd_pr_helper
./demo.sh
```

ArutyunovG commented:
This is a quick message, that I added a [Dockefile](https://github.com/ArutyunovG/pytorch_ssd_pr_helper/blob/master/Dockerfile) for that demo repo. It is simplifying reproduction even now, so decided to report.

I'm planning to add in this Dockerfile Caffe and TF and run their inference results with `demo.sh` to allow comparison with Caffe2.

In case you are going to actually use it, please note, that it requires [nvidia-docker)](https://github.com/NVIDIA/nvidia-docker)

ArutyunovG commented:
@BIT-silence 

FInished updates, which simplify checking the changes here.
These notes clarify the updates:

1. Added a [Dockerfile](https://github.com/ArutyunovG/pytorch_ssd_pr_helper/blob/master/Dockerfile).
What it does is
a) Download, build and install this PR's branch of Caffe2;
b) Download, build and install SSD branch of Caffe;
c) Install TensorFlow pip package;
d) Download the demo repo of this PR.
Please note, that it requires [nvidia-docker](https://github.com/NVIDIA/nvidia-docker)

2. Upgraded [demo.sh](https://github.com/ArutyunovG/pytorch_ssd_pr_helper/blob/master/demo.sh).
After the docker image is built one can compare the results between Caffe2 and reference frameworks inside the container. For that purpose, the needed commands after entering the container are:
```
cd pytorch_ssd_pr_helper
./demo.sh
```

Running the `demo.sh` file
1) Downloads SSD and FRCNN models in original framework as well as those same models converted to Caffe2
2) Runs Caffe2 inference on both of models
3) Runs Caffe inference on SSD model
4) Runs TF inference on FRCNN model
5) Compares the results with `epsilon == 0.01` 

To summarize: basically everything is automated now.

-------------

Hope this helps! :)

ArutyunovG commented:
@pytorchbot retest this please

ArutyunovG commented:
@pytorchbot retest this please

ArutyunovG commented:
This pull request was updated by merging new `legacy_plus_one` parameter.
It helps, the amount of changed files is 22, while it originally was 27! :)

To try this changes in dockefile (in case you use it), please rebuild it to have access to updated code. The caffe2 proto files of converted models were updated.

The not so good news are "`generate_proposals_gpu_test`" fails on rocm. It seems, something is wrong with the platform. The second test always fails there, if you try to add a new one (originally there is just one test). To demonstrate the issue, I intentioally moved the correct original test in the second place and now it fails. However, it is now in master and is obvously correct.

ArutyunovG committed:
TF OD API SSD and FRCNN support

ArutyunovG commented:
With new `legacy_plus_one` paramter, it seems, one can convert Faster R-CNNs from Gluon (MXNet) to Caffe2. But they also need some modification during conversion process to have a good look for production deployment.

Is example model interesting to share? I'm not sure about mxnet's `force_suppress` parameter yet, 
though. 

But results that we see are identical between MXNet and Caffe2 with `legacy_plus_one==0`. I.e. we can train now in mxnet as well, and still deploy in Caffe2. The same is true for SSDs from Gluon, if we have priorbox op.

ArutyunovG commented:
Is this alive?
It's a month since `legacy_plus_one` was added and two months since the last review. Probably, there is some internal work going on your side as with adding some new changes, that will allow the models to run, that I should not be aware of, but I just don't know if to drop support of this or not..

BIT-silence commented:
The implementation of this PR is a workable solution. While for both CPU and GPU version, we can do it better to consider other implementation method. I'm wondering if we can consider abandon this PR and then I will add the implementation from our side.

ArutyunovG commented:
Sure then I'm closing this and waiting for your solution. Thanks a lot for taking your time!

ArutyunovG commented:
Somehow, I deleted the models from dropbox with other files. They are restored now, you can use them from the demo repo if needed.
Just to clarify, the demo repo was [this one](https://github.com/ArutyunovG/pytorch_ssd_pr_helper).
